### DISCUSSION TITLE
How to make state global across clients

### DISCUSSION THREAD

**Discussion by jougs** [hyperdiv/hyperdiv#71] [Help]
Hi! Please excuse if I've overlooked this information throughout the documentation, but I haven't found anything about this topic, I thought I'd ask here:

(How) is it possible to synchronize state _across_ client connections?

My situation is like this: I'm building a simple UI for a IoT-style device based off a Raspberry Pi and this UI has a bunch of toggles and sliders that influence the state of some GPIO pins and PWM channels via a long-running background task. This part is working lovely so far. The problem shows, when a second client connects and sets controls differently. This state changes are not reflected in the UI of the other client and the GPIOs toggle between the states, which leads to flickering lights and is bad.

I've tried to apply a patterns like this:
```Python
state = None
def main():
    state = hd.state()
    hd.task().run(gpio_task, state)
```
However, while some state variables seem to respond on the other side, others don't and I fail to see how the ones differ from the others.

Is there some general advice for this kind of use case?

Many thanks!

**Comment by mariusnita** [comment]
Hi, Hyperdiv doesn't support this kind of shared state. All state is private to each client. I do plan to add an easy to use "shared reactive data" layer so all client UIs update when shared data is updated.

For now I only have clunky workarounds to suggest. Here's an example of an app with 2 sliders that derive their values from shared global data. Each client launches a task that watches a `threading.Event` and updates a piece of state, which causes that client to update. When one client changes the UI, the global data is updated and the other clients are notified to update. Note that this toy synchronization mechanism using a `threading.Event` is not robust.

```py
import threading
import hyperdiv as hd

# Shared global data
global_data = dict(x=0, y=1)

# A trivial synchronization mechanism.
evt = threading.Event()


def ui_update_task(state):
    while not state.stopped:
	evt.wait(timeout=1)
	state.update += 1
	evt.clear()


def main():
    with hd.box(padding=5, gap=1):
        x_slider = hd.slider(value=global_data["x"])
        y_slider = hd.slider(value=global_data["y"])

        # When a slider changes, update the global data, reset the
        # slider (so it keeps taking its value from the global data,
        # not its internal state), and notify the other clients to
        # refresh.
        if x_slider.changed:
            global_data["x"] = x_slider.value
            x_slider.reset()
            evt.set()

	if y_slider.changed:
            global_data["y"] = y_slider.value
            y_slider.reset()
            evt.set()

    state = hd.state(stopped=False, update=0)
    # Read state.update to create a read depenendency. So when this
    # state is mutated, the app updates.
    state.update

    lifecycle = hd.lifecycle()
    if lifecycle.app_stopped:
        # When the connection closes, tell the related task to exit.
	state.stopped = True

    # Launch the task.
    task = hd.task()
    task.run(ui_update_task, state)


hd.run(main)
```

**Comment by thedatadavis** [comment]
Thinking out loud here: this potentially gives me a reason to finally try
out [Tinybird](https://www.tinybird.co/).

The idea would be to post each state change to the events API and reload
from the workspace published endpoint (since it's real-time). I'm not quite
sure how to signal to other clients to refresh (outside of pulling latest
state after some action) but I can figure that out later.
